### ISSUE TITLE
window attention problem

### ISSUE THREAD

**Issue by gsfsdv** [DiT-3D/DiT-3D#14]
Hello, this is great work. I have some doubts about window attention. In the dit3d_window_atten.py code, window attention does not seem to reduce the dimension. Your code executes window attention first, but then reshape, isn't this equivalent to no execution?
![截图 2024-04-07 19-18-48](https://github.com/DiT-3D/DiT-3D/assets/19198171/54514885-607d-4c7d-8833-f598550dcfc7)

**Comment by sysuKaiYang** [comment]
I encountered the same problem. I found that when my configuration file uses --window_block_indexes '0,3,6,9', no matter which DiTBlock layer it is, its window_size variable is always 0, and it will never enter the line of code with window_partition; When I forcibly change the window_size to a non-zero value, an error will occur in the function of window_unpartition, simply because the 0th dimension changes back to B.

**Comment by EllingtonKirby** [comment]
Hi, I encountered a similar issue as the above. It seems there is a mismatch in the documentation between what the `window_unpartition` function expects and what is passed to it in the window attention `DiTBlock`. 

The `window_unpartition` expects input with the first dimensions as `B*num_windows`, as seen below.
![image](https://github.com/DiT-3D/DiT-3D/assets/8492979/b8fa4601-4ad9-477e-b1a4-20999c504d91)

But the input given to the function does not match this shape, as seen below. (the `attn` block results in the same `B, X, Y, Z, -1` shape)
![image](https://github.com/DiT-3D/DiT-3D/assets/8492979/2188588f-2c67-48b3-afc5-e3ed4343fbfa)

Given that the batch size is the actual first input to the `window_unpartition`, you can encounter the error seen by @sysuKaiYang for small batch sizes, which for me occurs at the end of my training loop.

I looked at the ViTDet paper but I did not see it explained why this dimension mismatch would be correct. 

Is this mis-match expected? How did you avoid such errors during your training? 

And if this mis-match is not expected, do you have a feeling over the correct way to resolve this error?

**Comment by P-Liseon** [comment]
if self.window_size > 0:
    X, Y, Z = x.shape[1], x.shape[2], x.shape[3]
    x, pad_xyz = window_partition(x, self.window_size)  # shape: (B*num_windows, W, W, W, C)
    x = x.reshape(-1, self.window_size**3, x.shape[-1])  # (B*num_windows, 64, C)
    x = modulate(x, shift_msa.repeat_interleave(8, dim=0), scale_msa.repeat_interleave(8, dim=0))
    x = x.reshape(-1, self.window_size, self.window_size, self.window_size, x.shape[-1])
    x = self.attn(x)
    x = window_unpartition(x, self.window_size, pad_xyz, (X, Y, Z))

窗口注意力计算的实现是否应该是这样，作者的实现真的做到了局部注意力的计算了吗？
